src/proxy/http/insert.rs

```diff
@@ -0,0 +1,226 @@
+{
+    type Service = Make<M, L, V>;
+
+    fn layer(&self, inner: M) -> Self::Service {
+        Self::Service {
+            inner,
+            lazy: self.lazy.clone(),
+            _marker: PhantomData,
+        }
+    }
+}
+
+// === impl Make ===
+
+impl<T, M, L, V> svc::Service<T> for Make<M, L, V>
+where
+    M: svc::Service<T>,
+    L: Lazy<V>,
+    V: Send + Sync + 'static,
+{
+    type Response = Service<M::Response, L, V>;
+    type Error = M::Error;
+    type Future = MakeFuture<M::Future, L, V>;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        self.inner.poll_ready()
+    }
+
+    fn call(&mut self, t: T) -> Self::Future {
+        Self::Future {
+            inner: self.inner.call(t),
+            lazy: self.lazy.clone(),
+            _marker: PhantomData,
+        }
+    }
+}
+
+// === impl MakeFuture ===
+
+impl<F, L, V> Future for MakeFuture<F, L, V>
+where
+    F: Future,
+    L: Lazy<V>,
+    V: Send + Sync + 'static,
+{
+    type Item = Service<F::Item, L, V>;
+    type Error = F::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let inner = try_ready!(self.inner.poll());
+        let svc = Service::new(inner, self.lazy.clone());
+        Ok(svc.into())
+    }
+}
+
+// === impl Service ===
+
+impl<S, L, V> Service<S, L, V> {
+    fn new(inner: S, lazy: L) -> Self {
+        Self {
+            inner,
+            lazy,
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<S, L, V, B> svc::Service<http::Request<B>> for Service<S, L, V>
+where
+    S: svc::Service<http::Request<B>>,
+    L: Lazy<V>,
+    V: Clone + Send + Sync + 'static,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = S::Future;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        self.inner.poll_ready()
+    }
+
+    fn call(&mut self, mut req: http::Request<B>) -> Self::Future {
+        req.extensions_mut().insert(self.lazy.value());
+        self.inner.call(req)
+    }
+}
+
+impl<V> Lazy<V> for ValLazy<V>
+where
+    V: Clone + Send + Sync + 'static,
+{
+    fn value(&self) -> V {
+        self.0.clone()
+    }
+}
+
+impl<F, V> Lazy<V> for FnLazy<F>
+where
+    F: Fn() -> V,
+    F: Clone,
+    V: Send + Sync + 'static,
+{
+    fn value(&self) -> V {
+        (self.0)()
+    }
+}
+
+pub mod target {
+    use super::*;
+
+    /// Wraps an HTTP `Service` so that the Stack's `T -typed target` is cloned into
+    /// each request's extensions.
+    #[derive(Clone, Debug)]
+    pub struct Make<M>(M);
+
+    pub struct MakeFuture<F, T> {
+        inner: F,
+        target: T,
+    }
+
+    // === impl Layer ===
+
+    pub fn layer<M>() -> impl svc::Layer<M, Service = Make<M>> + Copy {
+        svc::layer::mk(Make)
+    }
+
+    // === impl Stack ===
+
+    impl<T, M> svc::Service<T> for Make<M>
+    where
+        T: Clone + Send + Sync + 'static,
+        M: svc::Service<T>,
+    {
+        type Response = super::Service<M::Response, super::ValLazy<T>, T>;
```

Comment:
Not sure I'm following why the `Response` is a `super::Service` that takes a `lazy: ValLazy` instead of `lazy: FnLazy`. Is the use of `ValLazy` here mainly to work with the fact that `insert::layer()` takes a closure, and `insert::target::layer()` does not?

Reply:
Yep. target can't take a fn because it needs the ... target. it's probably not strictly necessary to have `ValLazy`